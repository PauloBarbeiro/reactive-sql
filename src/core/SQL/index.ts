import initSqlJs from 'sql.js'

export enum WasmSources {
    prod = 'https://cdnjs.cloudflare.com/ajax/libs/sql.js/1.6.1/sql-wasm.wasm',
    debug = 'https://cdnjs.cloudflare.com/ajax/libs/sql.js/1.6.1/sql-wasm-debug.wasm',
    test = 'node_modules/sql.js/dist/sql-wasm.wasm'
}

export type SqlValue = string | number | null | Uint8Array

export interface  QueryExecResult {
    columns: Array<string>
    values: Array<Array<SqlValue>>
}

export type BindParams = Array<SqlValue> | Record<string, SqlValue> | null

export interface SqlLite {
    exec: (query:string, params?: BindParams) => QueryExecResult
}

export interface DatabaseHolder {
    instance: SqlLite | null
    setInstance: (bd: SqlLite) => void
    getInstance: () => SqlLite | null
    destroy: () => void
}

export const database: DatabaseHolder = {
    instance: null,
    setInstance: function(db) {
        this.instance = db
    },
    getInstance: function() {
        return this.instance
    },
    destroy: function () {
        this.instance = null
    }
}

export const createSQL = async (path: string, name?: string) => {
    try{
        const SQL = await initSqlJs({
            locateFile: () => path
        })

        const db: SqlLite = new SQL.Database()
        database.setInstance(db)
        return db
    } catch (e) {
        console.error(e)
    }
    return
}

export const getDatabase = (): SqlLite | null => database.getInstance()

export const executeQuery = (query: string, params?: BindParams): QueryExecResult | undefined => {
    let db = getDatabase()

    if(!db) {
        console.error('SQL-lite instance not initiated! \nRun createSQL function to initialize the service.')
        return;
    }

    try {
        return db.exec(query, params)
    } catch (error) {
        console.error(`SQL-lite Error: ${error.message}`)
        return
    }
}

export default {
    createSQL,
    // getDatabase,
    executeQuery
}