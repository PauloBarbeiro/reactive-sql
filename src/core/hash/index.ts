const SEED = 5381;

const phash = (h: number, x: string): number => {
    let i = x.length;

    while (i) {
        h = (h * 33) ^ x.charCodeAt(--i);
    }

    return h;
};

// This is a djb2 hashing function
export const hash = (x: string): number => {
    return phash(SEED, x);
};
