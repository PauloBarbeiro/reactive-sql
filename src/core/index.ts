import { createSQL } from './SQL'

export default {
    createSQL
}