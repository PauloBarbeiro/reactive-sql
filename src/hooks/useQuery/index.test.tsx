import React from "react";
import {getByText, render, fireEvent, waitFor} from '@testing-library/react'
import { createSQL, WasmSources } from '../../core/SQL'
import {useQuery} from "./index";

// [{"columns": ["age", "name"], "values": [[18, "Paul"]]}]

const TestComponent = () => {
    const result = useQuery("SELECT age,name FROM test")

    const handleAddLennon = () => {
        console.log(' add Lennon')
    }

    const handleAddGeorge = () => {
        console.log(' add George')
    }

    if(!result || result.length === 0) {
        return (
            <p>No data available</p>
        )
    }

    const [data] = result

    return (
        <>
            <table>
                <caption>Table results</caption>
                <thead>
                <tr>
                    {data.columns.map((column, index) => (<th scope="col" key={index}>{column}</th>))}
                </tr>
                </thead>
                <tbody>
                {data.values.map(([age, name], index) => (
                    <tr key={index}>
                        <th scope="row">{name}</th>
                        <td>{age}</td>
                    </tr>
                ))}
                </tbody>
            </table>
            <button onClick={handleAddLennon}>Add Lennon</button>
            <button onClick={handleAddGeorge}>Add George</button>
        </>
    )
}

describe('useQuery', function () {
    beforeAll(async () => {
        const db = await createSQL(WasmSources.test)

        db!.exec(
            "DROP TABLE IF EXISTS test;\n"
            + "CREATE TABLE test (id INTEGER, age INTEGER, name TEXT);"
            + "INSERT INTO test VALUES ($id1, :age1, @name1);"
            + "INSERT INTO test VALUES ($id2, :age2, @name2);",
            {
                "$id1": 1, ":age1": 17, "@name1": "Ringo",
                "$id2": 2, ":age2": 18, "@name2": "Paul"
            }
        );
    })

    it('should execute a query and return some data', () => {
        const { container, getByText } = render(<TestComponent />)
        expect(container).toMatchSnapshot()
        expect(getByText('Ringo')).toBeTruthy()
        expect(getByText('Paul')).toBeTruthy()
    })

    it('should react to a change in the queried table', async () => {
        const { container, getByText } = render(<TestComponent />)
        expect(container).toMatchSnapshot()

        const addJohnBtn = getByText('Add Lennon')

        fireEvent.click(addJohnBtn)

        await waitFor(() => {
            expect(getByText('John')).toBeTruthy()
        })
    })
});
