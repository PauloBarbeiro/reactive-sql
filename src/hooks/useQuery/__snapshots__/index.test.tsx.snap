// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useQuery should execute a query and return some data 1`] = `
<div>
  <table>
    <caption>
      Table results
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          age
        </th>
        <th
          scope="col"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Ringo
        </th>
        <td>
          17
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Paul
        </th>
        <td>
          18
        </td>
      </tr>
    </tbody>
  </table>
  <button>
    Add Lennon
  </button>
  <button>
    Add George
  </button>
</div>
`;

exports[`useQuery should react to a change in the queried table 1`] = `
<div>
  <table>
    <caption>
      Table results
    </caption>
    <thead>
      <tr>
        <th
          scope="col"
        >
          age
        </th>
        <th
          scope="col"
        >
          name
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Ringo
        </th>
        <td>
          17
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Paul
        </th>
        <td>
          18
        </td>
      </tr>
    </tbody>
  </table>
  <button>
    Add John
  </button>
  <button>
    Add George
  </button>
</div>
`;
